src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,13 +620,120 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+    // below. This deals with both success and failure paths.
+    if (_last_replicate) {
+        auto fut = std::move(_last_replicate.value());
+        _last_replicate = std::nullopt;
+        if (!fut.available()) {
+            vlog(
+              _logger.debug, "Waiting for ongoing replication before syncing");
+
+            const auto before_replication = model::timeout_clock::now();
+
+            try {
+                std::ignore = co_await ss::with_timeout(
+                  before_replication + timeout, std::move(fut));
+            } catch (const ss::timed_out_error&) {
+                vlog(_logger.error, "Replication wait timed out");
+                co_return false;
+            }
+
+            const auto after_replication = model::timeout_clock::now();
+            timeout -= after_replication - before_replication;
+        } else {
+            co_await fut.discard_result();
+        }
+    }
+
+    if (!_raft->is_leader()) {
+        co_return false;
+    }
+
+    const auto last_applied_term = _insync_term;
+    const auto last_applied_offset = _manifest->get_insync_offset();
+    const auto current_term = _raft->term();
+    const auto committed_offset = _raft->committed_offset();
+
+    if (
+      last_applied_term == current_term
+      && last_applied_offset < committed_offset) {
+        vlog(
+          _logger.debug,
+          "Syncing archival STM from last applied offset {} to committed "
+          "offest {} in term {}",
+          last_applied_offset,
+          committed_offset,
+          current_term);
+
+        const auto deadline = model::timeout_clock::now() + timeout;
+        const bool synced = co_await wait_no_throw(committed_offset, deadline);
+
+        if (!synced) {
+            vlog(
+              _logger.warn,
+              "Failed to sync archival STM from offset {} to offest {} in "
+              "term {}",
+              last_applied_offset,
+              committed_offset,
+              current_term);
+        }
+
+        co_return synced;
+    } else {
+        vlog(
+          _logger.debug,
+          "Syncing archival STM to latest term {} from term {}",
+          current_term,
+          last_applied_term);
+        const bool synced = co_await persisted_stm::sync(timeout);
+
+        if (!synced) {
+            vlog(_logger.debug, "Failed to sync archival STM to latest term");
+        }
+
+        co_return synced;
+    }
+}
+
 ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
   model::record_batch batch, ss::abort_source& as) {
-    auto current_term = _insync_term;
-    auto fut = _raft->replicate(
-      current_term,
-      model::make_memory_record_batch_reader(std::move(batch)),
-      raft::replicate_options{raft::consistency_level::quorum_ack});
+    const auto current_term = _insync_term;
+
+    ss::promise<result<raft::replicate_result>> replication_promise;
+    _last_replicate = replication_promise.get_future();
```

Comment:
Maybe add assertion that checks that `_last_replicate` is equal to `std::nullopt`? We're calling `sync` every time before calling `do_relicate_commands` so the invariant `_last_replciate == std::nullopt` always holds but it's better to make it explicit.

Reply:
Another important thing is that previously, we could call `do_replicate_commands` independently from `sync`. The sync was called just in case we're at the start of the term I believe (don't quite remember to be honest). But now you have to cal `sync` before `do_replicate_commands` to clean `_last_replicate`. Otherwise it will barf after the promise is set. So maybe it makes sense to add some comment that explains this.

Reply:
Right. `sync` should be called before `do_replicate_commands` and both. calls should be under the stm lock.

Reply:
Added assertions to `do_replicate_commands`.